lib/i18n_sonder/workers/upload_source_strings_worker.rb

```diff
@@ -28,23 +29,89 @@ def perform(object_type, object_id, translated_attribute_params)
+          @logger.error("#{@log_pre} Can't find #{object_type} #{object_id} to send for translation")
         end
       end
 
       def handle_failure(exception)
         if exception.is_a? Exception
-          @logger.error("[I18nSonder::UploadSourceStringsWorker] #{exception}")
+          @logger.error("#{@log_pre} #{exception}")
+        end
+      end
+
+      private
+
+      # Look for duplicates of any of the +attributes_to_translate+, and update this model with valid duplicates.
+      # Filter down +attributes_to_translate+ to only the attribute-value pairs that don't have duplicate translations
+      # in all of the +I18nSonder.languages_to_translate+, since we do not need to upload those for translation.
+      def handle_duplicates(object, klass, attributes_to_translate)
+        locales_to_translate = I18nSonder.languages_to_translate.reject { |l| l == I18n.default_locale }.to_set
+
+        duplicates = duplicate_translations2(klass, attributes_to_translate)
+
+        duplicate_attrs_to_locales = attributes_to_translate.map { |a, _| [a, []] }.to_h
+
+        duplicates.each do |locale, translations|
+          unless translations.empty?
+            Mobility.with_locale(locale) do
+              @logger.info("#{@log_pre} Updating #{klass} #{object.id} with duplicate #{locale} translations for #{translations.keys}")
+              object.update(translations)
+            end
+          end
+
+          translations.each do |attr, _|
+            duplicate_attrs_to_locales[attr].append(locale)
+          end
+        end
+
+        attributes_to_translate.select do |attr, _|
+          locales_to_translate != duplicate_attrs_to_locales[attr].map(&:to_sym).to_set
+        end
+      end
+
+      # Look through all other models for translations of the same attribute-value pairs that need translation.
+      # Return those duplicates as a hash:
+      # { fr: { attribute1: "..." }, es: { attribute1: "..." }, ... }
+      #
+      # Valid translations are those where:
+      # 1) the attribute-values are equal for the model we are updating and the model we are comparing
+      # 2) the translation isn't stale (its updated time is after the translatable object's updated time)
+      #
+      # If there are multiple translations, take the one with the latest updated_at timestamp.
+      def duplicate_translations2(klass, attributes_to_translate)
+        attributes_to_translate.inject({}) do |translations_by_locale, (attr, value)|
+          duplicates = klass.joins("as k INNER JOIN #{translation_table_name(attr)} as t "\
+                    "ON t.translatable_id = k.id "\
+                    "AND t.translatable_type = '#{klass}' "\
+                    "AND t.key = '#{attr}'")
+                           .select("DISTINCT ON (t.locale) t.locale, t.value")
+                           .where("k.#{attr} = ?", value)
+                           .where("date_trunc('second', t.updated_at) >= date_trunc('second', k.updated_at)")
```

Comment:
is this the same effect as `t.updated_at >= k.updated_at`?

Reply:
Yes, same effect, it's just that the date can have milliseconds, which I didn't want to include in the comparison. There can be some millisecond level time differences since when we update a translation, since both the translation table and the model are getting their updated_at times changing at the same time.